File: backend/models/Report.js
const mongoose = require("mongoose");

const baseTransform = (doc, ret) => {
    ret.id = ret._id;
    delete ret._id;
    delete ret.__v;
    return ret;
};

const reportSchema = new mongoose.Schema(
    {
        owner_id: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "User", // FacilityOwner or PublicBody
            required: true,
        },
        title: {
            type: String,
            required: [true, "A saved report must have a title"],
            trim: true,
            maxlength: [100, "Report title cannot exceed 100 characters"],
        },
        notes: {
            type: String,
            trim: true,
            maxlength: [500, "Notes cannot exceed 500 characters"],
            default: "",
        },
    },
    {
        timestamps: true,
        discriminatorKey: "reportType",
        toJSON: {
            transform: baseTransform,
        },
    },
);

reportSchema.index({ owner_id: 1, createdAt: -1 });
reportSchema.index({ wall_id: 1 });

const Report = mongoose.model("Report", reportSchema);

// --- BASE REPORT ---
const baseReportSchema = new mongoose.Schema(
    {
        wall_id: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "Wall", // IndoorWall for Facility or OutdoorWall for PublicBody
            required: true,
        },
        reportData: {
            engagement: {
                totalSessions: Number,
                uniqueClimbers: Number,
                retentionRate: Number,
                avgTimeMins: Number,
                fastestTimeMins: Number,
                totalSends: Number,
                totalAttempts: Number,
            },
            quality: {
                avgRating: Number,
                totalReviews: Number,
                distribution: [
                    new mongoose.Schema(
                        { stars: Number, count: Number },
                        { _id: false },
                    ),
                ],
            },
            trends: {
                last30Days: [
                    new mongoose.Schema(
                        { date: String, sessions: Number },
                        { _id: false },
                    ),
                ],
                byDayOfWeek: [
                    new mongoose.Schema(
                        { day: Number, count: Number },
                        { _id: false },
                    ),
                ],
                byHourOfDay: [
                    new mongoose.Schema(
                        { hour: Number, count: Number },
                        { _id: false },
                    ),
                ],
            },
            // Qualitative feedback (Reviews)
            recentFeedback: [
                new mongoose.Schema(
                    { rating: Number, body: String, date: String },
                    { _id: false },
                ),
            ],
            // Qualitative feedback (Issues)
            recentIssues: [
                new mongoose.Schema(
                    { body: String, status: String, date: String },
                    { _id: false },
                ),
            ],
            // Demographic breakdown of climbers
            demographics: [
                new mongoose.Schema(
                    { bracket: String, count: Number },
                    { _id: false },
                ),
            ],
        },
    },
    {
        toJSON: {
            transform: baseTransform,
        },
    },
);

const BaseReport = Report.discriminator("BaseReport", baseReportSchema);

// --- GROUP REPORT ---
const groupReportSchema = new mongoose.Schema(
    {
        wall_ids: [
            {
                type: mongoose.Schema.Types.ObjectId,
                ref: "Wall", // IndoorWall for Facility or OutdoorWall for PublicBody
                required: true,
            },
        ],
        reportData: {
            /*! TODO: define the structure of aggregated report data for multiple walls.
             *  This could include averages, totals, and comparisons across walls.
             */
            aggregatedEngagement: {
                totalSessions: Number,
                uniqueClimbers: Number,
                retentionRate: Number,
                avgTimeMins: Number,
                fastestTimeMins: Number,
                totalSends: Number,
                totalAttempts: Number,
            },
            aggregatedQuality: {
                avgRating: Number,
                totalReviews: Number,
                distribution: [
                    new mongoose.Schema(
                        { stars: Number, count: Number },
                        { _id: false },
                    ),
                ],
            },
            aggregatedTrends: {
                last30Days: [
                    new mongoose.Schema(
                        { date: String, sessions: Number },
                        { _id: false },
                    ),
                ],
                byDayOfWeek: [
                    new mongoose.Schema(
                        { day: Number, count: Number },
                        { _id: false },
                    ),
                ],
                byHourOfDay: [
                    new mongoose.Schema(
                        { hour: Number, count: Number },
                        { _id: false },
                    ),
                ],
            },
            aggregatedFeedback: [
                new mongoose.Schema(
                    { rating: Number, body: String, date: String },
                    { _id: false },
                ),
            ],
            aggregatedIssues: [
                new mongoose.Schema(
                    { body: String, status: String, date: String },
                    { _id: false },
                ),
            ],

            aggregatedDemographics: [
                new mongoose.Schema(
                    { bracket: String, count: Number },
                    { _id: false },
                ),
            ],
            // Additional fields for group reports, such as comparisons between walls
            wallComparisons: [
                new mongoose.Schema(
                    {
                        wallId: {
                            type: mongoose.Schema.Types.ObjectId,
                            ref: "Wall",
                        },
                        engagement: {
                            totalSessions: Number,
                            uniqueClimbers: Number,
                            retentionRate: Number,
                            avgTimeMins: Number,
                            fastestTimeMins: Number,
                            totalSends: Number,
                            totalAttempts: Number,
                        },
                        quality: {
                            avgRating: Number,
                            totalReviews: Number,
                            distribution: [
                                new mongoose.Schema(
                                    { stars: Number, count: Number },
                                    { _id: false },
                                ),
                            ],
                        },
                    },
                    { _id: false },
                ),
            ],
        },
    },
    {
        toJSON: {
            transform: baseTransform,
        },
    },
);

const GroupReport = Report.discriminator("GroupReport", groupReportSchema);

module.exports = {
    Report,
    BaseReport,
    GroupReport,
};

File: frontend/lib/models/report.dart
class Report {
  final String id;
  final String title;
  final String notes;
  final DateTime createdAt;
  final Map<String, dynamic>? wall;
  final List<Map<String, dynamic>>? walls;
  final ReportData? reportData;

  Report({
    required this.id,
    required this.title,
    required this.notes,
    required this.createdAt,
    this.wall,
    this.walls,
    this.reportData,
  });

  factory Report.fromJson(Map<String, dynamic> json) {
    final createdAtValue = json['createdAt'];
    final parsedCreatedAt = createdAtValue is String
        ? DateTime.tryParse(createdAtValue) ??
              DateTime.fromMillisecondsSinceEpoch(0)
        : createdAtValue is int
        ? DateTime.fromMillisecondsSinceEpoch(createdAtValue)
        : DateTime.fromMillisecondsSinceEpoch(0);

    final wallList = (json['wall_ids'] ?? json['walls'] ?? []) as List;

    return Report(
      id: json['_id'] ?? json['id'] ?? '',
      title: json['title'] ?? '',
      notes: json['notes'] ?? '',
      createdAt: parsedCreatedAt,
      wall: json['wall_id'] is Map
          ? Map<String, dynamic>.from(json['wall_id'])
          : null,
      walls: wallList
          .whereType<Map>()
          .map((entry) => Map<String, dynamic>.from(entry))
          .toList(),
      reportData: json['reportData'] != null
          ? ReportData.fromJson(json['reportData'])
          : null,
    );
  }
}

class ReportData {
  final String? wallId;
  final String? wallName;
  final Map<String, dynamic> engagement;
  final Map<String, dynamic> quality;
  final List<dynamic> trends;

  final List<dynamic> byDayOfWeek;
  final List<dynamic> byHourOfDay;
  final List<dynamic> recentFeedback;
  final List<dynamic> demographics;
  final List<dynamic> recentIssues;
  final List<Map<String, dynamic>> wallComparisons;

  ReportData({
    this.wallId,
    this.wallName,
    required this.engagement,
    required this.quality,
    required this.trends,
    required this.byDayOfWeek,
    required this.byHourOfDay,
    required this.recentFeedback,
    required this.demographics,
    required this.recentIssues,
    required this.wallComparisons,
  });

  factory ReportData.fromJson(Map<String, dynamic> json) {
    final aggregatedEngagement =
        json['aggregatedEngagement'] ?? json['engagement'] ?? {};
    final aggregatedQuality =
        json['aggregatedQuality'] ?? json['quality'] ?? {};
    final aggregatedTrends = json['aggregatedTrends'] ?? json['trends'] ?? {};

    return ReportData(
      wallId: json['wallId']?.toString(),
      wallName: json['wallName']?.toString(),
      engagement: aggregatedEngagement,
      quality: aggregatedQuality,
      trends:
          aggregatedTrends['last30Days'] ?? aggregatedTrends['trends'] ?? [],
      byDayOfWeek: aggregatedTrends['byDayOfWeek'] ?? [],
      byHourOfDay: aggregatedTrends['byHourOfDay'] ?? [],
      recentFeedback:
          json['aggregatedFeedback'] ?? json['recentFeedback'] ?? [],
      demographics:
          json['aggregatedDemographics'] ?? json['demographics'] ?? [],
      recentIssues: json['aggregatedIssues'] ?? json['recentIssues'] ?? [],
      wallComparisons: (json['wallComparisons'] as List<dynamic>? ?? [])
          .map((item) => Map<String, dynamic>.from(item as Map))
          .toList(),
    );
  }
}
